### PR TITLE
enable deferred cluster creation

### PR DESCRIPTION
**Pull request by ramaraochavali**
Enables deferred cluster creation on worker threads

- [ ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Dual Stack
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

### CODE REVIEW

**Review by keithmattix — COMMENTED** [review]
What's the motivation for this?

**Review by howardjohn — COMMENTED** [review]
same question, when I tried in https://github.com/istio/istio/issues/49868 I saw no benefits

**Comment by ramaraochavali** [comment]
I am seeing improvement in server initialization times with two static clusters and concurrency of 50

server.initialization_time_ms: P0(nan,100) P25(nan,102.5) P50(nan,105) P75(nan,107.5) P90(nan,109) P95(nan,109.5) P99(nan,109.9) P99.5(nan,109.95) P99.9(nan,109.99) P100(nan,110)

server.initialization_time_ms: P0(nan,140) P25(nan,142.5) P50(nan,145) P75(nan,147.5) P90(nan,149) P95(nan,149.5) P99(nan,149.9) P99.5(nan,149.95) P99.9(nan,149.99) P100(nan,150)

Will do more tests and report back

**Comment by ramaraochavali** [comment]
Here are the results of my test

**Concurrency 2 Clusters 3600**

**Without Deferred creation**

server.memory_allocated: 759852680
server.memory_heap_size: 1067294720
server.memory_physical_size: 834822144

server.initialization_time_ms: P0(nan,46000) P25(nan,46250) P50(nan,46500) P75(nan,46750) P90(nan,46900) P95(nan,46950) P99(nan,46990) P99.5(nan,46995) P99.9(nan,46999) P100(nan,47000)

**With Deferred creation**

server.memory_allocated: 749908528
server.memory_heap_size: 1056219136
server.memory_physical_size: 853909504

server.initialization_time_ms: P0(nan,36000) P25(nan,36250) P50(nan,36500) P75(nan,36750) P90(nan,36900) P95(nan,36950) P99(nan,36990) P99.5(nan,36995) P99.9(nan,36999) P100(nan,37000)


**Concurrency 10 Clusters 3600**

**Without Deferred creation**

server.memory_allocated: 901796592
server.memory_heap_size: 1266057216
server.memory_physical_size: 981704704

server.initialization_time_ms: P0(nan,52000) P25(nan,52250) P50(nan,52500) P75(nan,52750) P90(nan,52900) P95(nan,52950) P99(nan,52990) P99.5(nan,52995) P99.9(nan,52999) P100(nan,53000)


**With**

server.memory_allocated: 783817120
server.memory_heap_size: 1104920576
server.memory_physical_size: 858816512

server.initialization_time_ms: P0(nan,30000) P25(nan,30250) P50(nan,30500) P75(nan,30750) P90(nan,30900) P95(nan,30950) P99(nan,30990) P99.5(nan,30995) P99.9(nan,30999) P100(nan,31000)


**Summary:**

We see improvement both in memory and init times.  Out of these 3600 clusters, 400 clusters STRICT_DNS type of clusters and deferred creation does not work for them.

**Comment by ramaraochavali** [comment]
@howardjohn  can you please share /stats?usedonly endpoint output of your test when you saw no difference? Please enable all stats of Envoy.

**Comment by hzxuzhonghu** [comment]
please wait if possible. i want to take a look,but i am on holiday this week

**Comment by ramaraochavali** [comment]
sure it is not urgent

**Comment by hzxuzhonghu** [comment]
How does this work in envoy, will it ceate cluster on flight, so will it influence the latency?

**Comment by ramaraochavali** [comment]
Yes. When a request comes on the worker thread, it creates it inline with the already existing object https://github.com/envoyproxy/envoy/blob/8d4cf00e2980aeba11b9b74230a53f92086419bc/source/common/upstream/cluster_manager_impl.cc#L1355. Should not incur lot of latency

**Review by costinm — CHANGES_REQUESTED** [review]
I like this change - and any 'on demand'/'delayed' or scale improving PR - but absolutely not as a default behavior change. 

Add an annotation or option or meta - off by default - and it would be great. Gradually we can make it the default - maybe for waypoints first ( no backward compat risks), some classes of gateways - and maybe even sidecars, but over few releases.

**Comment by kyessenov** [comment]
This needs to be gated by a feature flag and reversible. Otherwise, this is probably a good change.

**Comment by ramaraochavali** [comment]
@kyessenov @costinm added feature flag. PTAL

@howardjohn when you get chance, please share stats of your experiment

**Comment by howardjohn** [comment]
I jsut have https://github.com/istio/istio/issues/49868#issuecomment-1994591948

File: releasenotes/notes/deferred_cluster_creation.yaml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+apiVersion: release-notes/v2
+kind: feature
+area: traffic-management
+
+releaseNotes:
+- |
+  **Added** an experimental feature to enable cluster creation on worker threads inline during requests.
+  This will save memory and CPU cycles in cases where there are lots of inactive clusters and > 1 worker thread.
+  This can be enabled by setting ENABLE_DEFERRED_CLUSTER_CREATION to true in Istiod Deployment.
```
Comment (howardjohn) [inline releasenotes/notes/deferred_cluster_creation.yaml:9]: This is actually in the agent not istiod?

**Review by howardjohn — APPROVED** [review]
LGTM. I am fine with turning it on by default with "turn off for compatibility version". Or we can get more testing then flip it

File: releasenotes/notes/deferred_cluster_creation.yaml (hunk 1 above)
Comment (hzxuzhonghu) [reply releasenotes/notes/deferred_cluster_creation.yaml:9]: Then the feature flag should have its own home not under pilot

**Review by hzxuzhonghu — COMMENTED** [review]
any latency test result

File: releasenotes/notes/deferred_cluster_creation.yaml (hunk 1 above)
Comment (ramaraochavali) [reply releasenotes/notes/deferred_cluster_creation.yaml:9]: Oh yes. I forgot just copy pasted some other flag :-)

> Then the feature flag should have its own home not under pilot

We will have to do that refactor

**Comment by ramaraochavali** [comment]
> any latency test result

latency is very negligible based on my measurements

**Comment by ramaraochavali** [comment]
compatibility check added. 

@costinm  can you please review?

File: releasenotes/notes/deferred_cluster_creation.yaml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+apiVersion: release-notes/v2
+kind: feature
+area: traffic-management
+
+releaseNotes:
+- |
+  **Added** an experimental feature to enable cluster creation on worker threads inline during requests.
+  This will save memory and CPU cycles in cases where there are lots of inactive clusters and > 1 worker thread.
+  This can be disabled by setting ENABLE_DEFERRED_CLUSTER_CREATION to true in agent Deployment.
```
Comment (wulianglongrd) [inline releasenotes/notes/deferred_cluster_creation.yaml:9]: `ENABLE_DEFERRED_CLUSTER_CREATION` starts with `enable`, but `true` means `disable`? This confuses me.

Comment (ramaraochavali) [reply releasenotes/notes/deferred_cluster_creation.yaml]: sorry. good point, That needs to change after the default has been changed. fixed now

File: releasenotes/notes/deferred_cluster_creation.yaml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+apiVersion: release-notes/v2
+kind: feature
+area: traffic-management
+
+releaseNotes:
+- |
+  **Added** an experimental feature to enable cluster creation on worker threads inline during requests.
+  This will save memory and CPU cycles in cases where there are lots of inactive clusters and > 1 worker thread.
+  This can be disabled by setting ENABLE_DEFERRED_CLUSTER_CREATION to false in agent Deployment.
```
Comment (hzxuzhonghu) [inline releasenotes/notes/deferred_cluster_creation.yaml:9]: need update, now it is not enabled by default

**Comment by hzxuzhonghu** [comment]
>latency is very negligible based on my measurements

I doubt this conclusion

**Comment by ramaraochavali** [comment]
Please see what Envoy is doing when creating cluster in line https://github.com/envoyproxy/envoy/blob/8d4cf00e2980aeba11b9b74230a53f92086419bc/source/common/upstream/cluster_manager_impl.cc#L1355 - I do not think it is doing any significant work on a worker. But feel free to measure with your service

**Comment by ramaraochavali** [comment]
@costinm gentle ping

**Review by costinm — APPROVED** [review]

**Pull request merged**